FLT/Mathlib/NumberTheory/Padics/PadicIntegers.lean

```diff
@@ -51,23 +52,124 @@ lemma isOpenEmbedding_coe : IsOpenEmbedding ((↑) : ℤ_[p] → ℚ_[p]) := by
 -- Yaël: Do we really want this as a coercion?
 noncomputable instance : Coe ℤ_[p]ˣ ℚ_[p]ˣ where coe := Units.map Coe.ringHom.toMonoidHom
 
-/-- For a `ℤ_[p]`-submodule `s` of `ℚ_[p]`, `x • s` has index `‖x‖⁻¹` in `s`.
+/-- `x • S` has index `‖x‖⁻¹` in `S`, where `S` is the copy of `ℤ_[p]` inside `ℚ_[p]` --/
+
+lemma smul_submodule_one_relindex (hx: x ≠ 0):
+    (x • (1 : Submodule ℤ_[p] ℚ_[p]).toAddSubgroup).relindex (1 : Submodule ℤ_[p] ℚ_[p]).toAddSubgroup = ‖x.val‖₊⁻¹ := by
+
+  -- We first prove a closely related statement using subgroups of `ℤ_[p]`. We then use the coercion from
+  -- ℤ_[p] to ℚ_[p] to obtain the desired result for the groups considered
+  -- as subgroups of `ℚ_[p]`.
+
+  let H := (x • (1 : Submodule ℤ_[p] ℤ_[p])).toAddSubgroup
+
+  have H_relindex_Z : (H.index) = ‖(x : ℚ_[p])‖₊⁻¹ := by
+    have x_factor := PadicInt.unitCoeff_spec hx
+    norm_cast at x_factor
+    unfold H
+    nth_rw 1 [x_factor]
+    conv =>
+      pattern _ • _
+      equals (Ideal.span {(p: ℤ_[p]) ^ (x.valuation)}) =>
+        ext y
+        refine ⟨?_, ?_⟩
```

Comment:
I think the case split here with `ext` and `refine` is an inefficient approach. Your goal is `(u * r) • 1 = Ideal.span {r}` so this should just be two rewrites: firstly `t • 1 = Ideal.span {t}` and secondly `Ideal.span {u * r} = Ideal.span {r}` for `u` a unit, and both of these should be lemmas in mathlib (and if they're not, then just make those lemmas here). Again I really want to encourage you to break down your big proof into many smaller proofs. My two-line proof is the abstraction you want, you don't want to be grinding out 15-line proofs, it doesn't scale. Note also that your `norm_cast` on line 68 now looks like a step in the wrong direction.